substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -982,43 +996,69 @@ impl<T: Config> Pallet<T> {
 	///
 	/// If the nominator already exists, their nominations will be updated.
 	///
-	/// NOTE: you must ALWAYS use this function to add nominator or update their targets. Any access
-	/// to `Nominators` or `VoterList` outside of this function is almost certainly
-	/// wrong.
+	/// NOTE: you must ALWAYS use this function to add nominator or update their targets. Any
+	/// access to `Nominators` or `VoterList` outside of this function is almost certainly wrong.
 	pub fn do_add_nominator(who: &T::AccountId, nominations: Nominations<T>) {
-		if !Nominators::<T>::contains_key(who) {
-			// maybe update sorted list.
-			let _ = T::VoterList::on_insert(who.clone(), Self::weight_of(who))
-				.defensive_unwrap_or_default();
-		}
-		Nominators::<T>::insert(who, nominations);
+		let nomination_accounts = nominations.targets.to_vec();
+
+		match (Nominators::<T>::contains_key(who), T::VoterList::contains(who)) {
+			(false, false) => {
+				// new nomination
+				Nominators::<T>::insert(who, nominations);
+				T::EventListeners::on_nominator_add(who, nomination_accounts);
+			},
+			(_, true) => {
+				// update nominations or un-chill nominator.
+				let prev_nominations = Self::nominations(who).unwrap_or_default();
+				Nominators::<T>::insert(who, nominations);
+				T::EventListeners::on_nominator_update(who, prev_nominations, nomination_accounts);
+			},
+			(true, false) => {
+				defensive!("unexpected state.");
+			},
+		};
 
 		debug_assert_eq!(
 			Nominators::<T>::count() + Validators::<T>::count(),
-			T::VoterList::count()
+			T::VoterList::iter()
+				.filter(|v| Self::status(&v) != Ok(StakerStatus::Idle))
+				.count() as u32,
 		);
 	}
 
+	/// Tries to chill a nominator.
+	///
+	/// A chilled nominator is removed from the `Nominators` map, and the nominator's new state must
+	/// be signalled to [`T::EventListeners`].
+	pub(crate) fn do_chill_nominator(who: &T::AccountId) -> bool {
+		Nominators::<T>::take(who).map_or(false, |nominations| {
+			T::EventListeners::on_nominator_remove(who, nominations.clone().targets.into());
+			true
+		})
```

Comment:
nit: This would be cleaner with a `match` or simply `map` followed by `is_some`.